### PR TITLE
Update Anchor EA to new way of deriving USD/bETH price price

### PR DESCRIPTION
**Pull request by cds95**
## Description

This PR updates the way the `USD/bETH` price is derived.  The high level calculation is outlined below

```
bETH = ETHPrice * StETHETHRate / StETHbETHExchangeRate
```

ETHPrice: Pulled from a the token allocation EA
StETHETHRate: `ETH / stETH` price on Curve - Pulled from curve smart contract
StETHbETHExchangeRate - pulled from Lido's AnchorVault contract by calling get_rate() 

## Changes

 - Update composite Anchor EA to calculate the `bETH` price using the method above
 - Update tests to reflect changes
 - Add new environment variable `STETH_POOL_CONTRACT_ADDRESS` for the `stETH-ETH` pool in Curve with a default of `0xdc24316b9ae028f1497c275eb9192a3ea0f67022` 

## Steps to Test

1. Set the Token Allocation EA required environment variables
2. Run the EA
3. Verify that the EA returns a successful response with the following sources

* tiingo
* coinmetrics
* cryptoompare
* ncfx
* cfbenchmarks
* finage

Sample request

```
{
  "id": "1",
  "data": {
    "from": "BETH",
    "to": "USD",
    "source": "ncfx"
  }
}
```

## Quality Assurance

- [x] Ran `yarn changeset` if adapter source code was changed
- [x] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `<ADAPTER_PACKAGE>/schemas/env.json` and `<ADAPTER_PACKAGE>/README.md`
- [x] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `infra-k8s` configuration file.
- [x] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Clubhouse/Shortcut
- [x] This is related to a maximum of one Clubhouse/Shortcut story or GitHub issue
- [x] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types)

### CODE REVIEW

File: packages/composites/anchor/src/endpoint/price/beth.ts [hunk 1]
```diff
@@ -41,3 +25,33 @@ export const execute: PriceExecute = async (input, _, config, taAdapterResponse)
...
+  const anchorVaultContract = new ethers.Contract(
+    config.anchorVaultContractAddress,
+    anchorVaultAbi,
+    provider,
+  )
+  const bEthExchangeRateBigNum = await anchorVaultContract.get_rate()
+  // Need to convert to BigNumber JS from ethers BigNumber as the latter will remove all decimals
+  return new BigNumber(bEthExchangeRateBigNum.toString())
+    .dividedBy(new BigNumber(10).pow(18))
+    .toNumber()
```
Comment (boxhock) [inline packages/composites/anchor/src/endpoint/price/beth.ts:56]: Why are we converting this to a number, which would lose precision?

File: packages/composites/anchor/src/endpoint/price/beth.ts [hunk 2]
```diff
@@ -41,3 +25,33 @@ export const execute: PriceExecute = async (input, _, config, taAdapterResponse)
...
+  provider: ethers.providers.JsonRpcProvider,
+): Promise<number> => {
+  const stEthPoolContract = new ethers.Contract(
+    config.stEthPoolContractAddress,
+    curvePoolAbi,
+    provider,
+  )
+  const ethBal = await stEthPoolContract.balances(0)
+  const stEthBal = await stEthPoolContract.balances(1)
+  return new BigNumber(ethBal.toString()).dividedBy(new BigNumber(stEthBal.toString())).toNumber()
```
Comment (boxhock) [inline packages/composites/anchor/src/endpoint/price/beth.ts:40]: Recommend not making this to a number, which loses precision.

File: packages/composites/anchor/src/endpoint/price/beth.ts [hunk 3]
```diff
@@ -1,37 +1,21 @@
...
-  // Need to convert to BigNumber JS from ethers BigNumber as the latter will remove all decimals
-  const stEthPerBEth = new BigNumber(bEthExchangeRateBigNum.toString())
-    .dividedBy(new BigNumber(10).pow(18))
-    .toNumber()
-  const usdPerStEth = taAdapterResponse.data.result
-  const result = usdPerStEth * stEthPerBEth
+  const stEthPerBEth = await getStEthBEthExchangeRate(config, provider)
+  const ethPerStEth = await getStETHExchangeRate(config, provider)
+  const usdPerEth = taAdapterResponse.data.result
+  const result = usdPerEth * ethPerStEth * stEthPerBEth
```
Comment (boxhock) [inline packages/composites/anchor/src/endpoint/price/beth.ts:18]: Should probably be multiplied with ethers.BigNumber

File: packages/composites/anchor/src/utils.ts [hunk 4]
```diff
@@ -14,7 +14,7 @@ export const getTokenPrice = async (
   const allocations = [
     {
       symbol,
-      balance: BigNumber.from(10).pow(decimals),
+      balance: BigNumber.from(10).pow(decimals).toString(),
```
Comment (boxhock) [inline packages/composites/anchor/src/utils.ts:17]: 👍

**Review by boxhock — CHANGES_REQUESTED** [review]
Mainly concerns about losing precision by turning BigNumbers into native numbers/float

File: packages/composites/anchor/src/endpoint/price/beth.ts (hunk 1 above)
Comment (cds95) [reply packages/composites/anchor/src/endpoint/price/beth.ts:56]: So I turned this to a number as the result from token allocations is returned as number and they are multiplied together later on. Do you recommend converting the TA response to a big number first?

Comment (cds95) [reply packages/composites/anchor/src/endpoint/price/beth.ts]: decided to use BigNumber

File: packages/composites/anchor/src/endpoint/price/beth.ts [hunk 5]
```diff
@@ -1,37 +1,23 @@
...
-    .toNumber()
-  const usdPerStEth = taAdapterResponse.data.result
-  const result = usdPerStEth * stEthPerBEth
+  const stEthPerBEth = await getStEthBEthExchangeRate(config, provider)
+  const ethPerStEth = await getStETHExchangeRate(config, provider)
+  const usdPerEth = taAdapterResponse.data.result
+  const result = new BigNumber(usdPerEth)
+    .multipliedBy(ethPerStEth)
+    .multipliedBy(stEthPerBEth)
+    .toString()
```
[the remark this answers is not in the document]
Comment (cds95) [reply packages/composites/anchor/src/endpoint/price/beth.ts]: gotcha ok just changed this.

**Review by boxhock — APPROVED** [review]

**Pull request merged**